### PR TITLE
Init interchain contracts

### PR DESCRIPTION
**Pull request by ChiTimesChi**
**Description**
A clear and concise description of the features you're adding in this pull request.

**Additional context**
Add any other context about the problem you're solving.

**Metadata**
- Fixes #[Link to Issue]

## Summary by CodeRabbit

- **New Features**
    - Introduced abstract and locking processors for converting between ERC20 and ICERC20 tokens.
- **Tests**
    - Added test cases for the locking processor contract.
- **Chores**
    - Updated CI/CD pipelines and configurations for the new contracts-interchain package.

### CODE REVIEW

File: packages/contracts-interchain/src/libs/RateLimit.sol [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            return self.totalLimit;
+        }
+        // Otherwise, figure out how much the limit has replenished since the last update
+        uint256 replenished = (timePassed * self.totalLimit) / REPLENISH_PERIOD;
+        limit = self.lastRemaining + replenished;
+        // Limit cannot surpass the total limit
+        if (limit > self.totalLimit) {
+            limit = self.totalLimit;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/libs/RateLimit.sol:81]: Dismissed: Time based rate limiting is supposed to take into account `block.timestamp`

File: packages/contracts-interchain/src/tokens/InterchainERC20.sol [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    /// @dev Rate Limit for Bridge's burn operations
+    mapping(address bridge => RateLimit) internal _burnLimits;
+    /// @dev Rate Limit for Bridge's mint operations
+    mapping(address bridge => RateLimit) internal _mintLimits;
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        address initialAdmin_,
+        address processor_
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/tokens/InterchainERC20.sol]: Dismissed: the processor address could be a zero address on the chains without the underlying token.

File: packages/contracts-interchain/src/libs/Create2.sol [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    function deploy(uint256 value, bytes32 salt, bytes memory creationCode) internal returns (address deployedAt) {
+        deployedAt = predictDeployment(address(this), salt, creationCode);
+        if (deployedAt.code.length > 0) revert Create2__AlreadyDeployed(deployedAt);
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            // create2(value, offset, size, salt)
+            deployedAt := create2(value, add(creationCode, 0x20), mload(creationCode), salt)
+        }
+        if (deployedAt == address(0)) revert Create2__DeploymentFailed();
+    }
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/libs/Create2.sol:24]: Dismissed: use of assembly is required here for for explicit CREATE2 behavior.

File: packages/contracts-interchain/src/tokens/InterchainERC20.sol [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    /// @dev Rate Limit for Bridge's burn operations
+    mapping(address bridge => RateLimit) internal _burnLimits;
+    /// @dev Rate Limit for Bridge's mint operations
+    mapping(address bridge => RateLimit) internal _mintLimits;
+
+    error InterchainERC20__AdminZero();
+
+    constructor(string memory name_, string memory symbol_, uint8 decimals_) ERC20(name_, symbol_) {
+        (address initialAdmin, address processor) = IInterchainFactory(msg.sender).getInterchainTokenDeployParameters();
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/tokens/InterchainERC20.sol:32]: Dismissed: the processor address could be a zero address on the chains without the underlying token.

File: packages/contracts-interchain/src/factories/InterchainFactory.sol [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        // Deploy InterchainERC20. We use assert statement as a sanity check for create2 prediction.
+        address deployment = _deployContract(tokenCreationCode, salt, initialAdmin, processor);
+        assert(deployment == interchainToken);
+        emit InterchainTokenDeployed(interchainToken, underlyingToken);
+        // Deploy Processor. We use assert statement as a sanity check for create2 prediction.
+        // Note: Processor contract doesn't require any constructor arguments.
+        deployment = _deployContract(processorCreationCode, salt, interchainToken, underlyingToken);
+        assert(deployment == processor);
+        emit ProcessorDeployed(processor);
+    }
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/factories/InterchainFactory.sol:75]: Dismissed: tight packing is required here to determine the creation code for a contract with constructor arguments.

File: packages/contracts-interchain/src/factories/InterchainFactory.sol [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        bytes memory tokenCode
+    )
+        external
+        view
+        returns (address)
+    {
+        bytes32 salt = _getDeploymentSalt(deployer, address(0));
+        bytes memory tokenCreationCode = abi.encodePacked(tokenCode, abi.encode(name, symbol, decimals));
+        return Create2.predictDeployment(address(this), salt, tokenCreationCode);
+    }
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/factories/InterchainFactory.sol:134]: Dismissed: tight packing is required here to determine the creation code for a contract with constructor arguments.

File: packages/contracts-interchain/src/factories/InterchainFactory.sol [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        bytes memory tokenCode
+    )
+        external
+        view
+        returns (address)
+    {
+        bytes32 salt = _getDeploymentSalt(deployer, underlyingToken);
+        bytes memory tokenCreationCode = abi.encodePacked(tokenCode, _deriveMetadata(underlyingToken));
+        return Create2.predictDeployment(address(this), salt, tokenCreationCode);
+    }
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/factories/InterchainFactory.sol:112]: Dismissed: tight packing is required here to determine the creation code for a contract with constructor arguments.

File: packages/contracts-interchain/src/tokens/InterchainERC20.sol [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    bytes32 public constant EMERGENCY_PAUSER_ROLE = keccak256("EMERGENCY_PAUSER_ROLE");
+    bytes32 public constant GOVERNOR_ROLE = keccak256("GOVERNOR_ROLE");
+
+    /// @notice Address of the processor contract, responsible for Token<>InterchainToken conversions
+    /// Note: this should be left as zero address, if the InterchainERC20 is used as a "native token deployment"
+    /// on the chain rather than an interchain representation of an existing token.
+    address public immutable PROCESSOR;
+
+    /// @dev Number of decimals for the token, setup during the contract deployment
+    uint8 internal immutable _DECIMALS;
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/tokens/InterchainERC20.sol:22]: Dismissed: immutable variables should be using the same case as constants.

File: packages/contracts-interchain/src/tokens/InterchainERC20.sol [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+import {Pausable} from "@openzeppelin/contracts/utils/Pausable.sol";
+
+contract InterchainERC20 is ERC20, AccessControl, Pausable, ICERC20 {
+    bytes32 public constant EMERGENCY_PAUSER_ROLE = keccak256("EMERGENCY_PAUSER_ROLE");
+    bytes32 public constant GOVERNOR_ROLE = keccak256("GOVERNOR_ROLE");
+
+    /// @notice Address of the processor contract, responsible for Token<>InterchainToken conversions
+    /// Note: this should be left as zero address, if the InterchainERC20 is used as a "native token deployment"
+    /// on the chain rather than an interchain representation of an existing token.
+    address public immutable PROCESSOR;
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/tokens/InterchainERC20.sol:19]: Dismissed: immutable variables should be using the same case as constants.

File: packages/contracts-interchain/src/processors/AbstractProcessor.sol [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+/// while the Processor is simply a tool for Token<>InterchainToken conversion.
+/// Therefore, the minting/burning of both tokens does not require a separate
+/// minting/burning limit for the Processor.
+/// The AbstractProcessor implements the IDefaultPool interface, allowing a seamless integration
+/// with the SynapseBridge contract.
+abstract contract AbstractProcessor is IDefaultPool {
+    using SafeERC20 for IERC20;
+
+    address public immutable INTERCHAIN_TOKEN;
+    address public immutable UNDERLYING_TOKEN;
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/processors/AbstractProcessor.sol:22]: Dismissed: immutable variables should be using the same case as constants.

File: packages/contracts-interchain/src/processors/AbstractProcessor.sol [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+/// NOTE: ICERC20 token issuance by the Bridge is rate limited by burn and mint limits,
+/// while the Processor is simply a tool for Token<>InterchainToken conversion.
+/// Therefore, the minting/burning of both tokens does not require a separate
+/// minting/burning limit for the Processor.
+/// The AbstractProcessor implements the IDefaultPool interface, allowing a seamless integration
+/// with the SynapseBridge contract.
+abstract contract AbstractProcessor is IDefaultPool {
+    using SafeERC20 for IERC20;
+
+    address public immutable INTERCHAIN_TOKEN;
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/processors/AbstractProcessor.sol:21]: Dismissed: immutable variables should be using the same case as constants.

File: packages/contracts-interchain/src/tokens/InterchainERC20.sol [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/tokens/InterchainERC20.sol:19]: Dismissed: immutable variables should be using the same case as constants.

File: packages/contracts-interchain/src/factories/InterchainFactory.sol [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+    )
+        external
+    {
+        // Use underlyingToken = address(0) to deploy a standalone InterchainERC20
+        bytes32 salt = _getDeploymentSalt(msg.sender, address(0));
+        bytes memory tokenCreationCode = abi.encodePacked(tokenCode, abi.encode(name, symbol, decimals));
+        // Use processor = address(0) to deploy a standalone InterchainERC20
+        address interchainToken = _deployContract(tokenCreationCode, salt, initialAdmin, address(0));
+        emit InterchainTokenDeployed(interchainToken, address(0));
+    }
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/factories/InterchainFactory.sol:44]: This is not required, as it is the token responsibility to check the parameters during construction.

File: packages/contracts-interchain/src/factories/InterchainFactory.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/factories/InterchainFactory.sol:75]: This is not required, as it is the token and processor responsibility to check the parameters during construction.

File: packages/contracts-interchain/src/factories/InterchainFactory.sol [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+    )
+        internal
+        returns (address deployedAt)
+    {
+        _firstArg = firstArg;
+        _secondArg = secondArg;
+        deployedAt = Create2.deploy(0, salt, creationCode);
+        delete _firstArg;
+        delete _secondArg;
+    }
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-interchain/src/factories/InterchainFactory.sol:168]: This check is implemented in the Create2 library, thus not required in the Factory contract.

**Comment by trajan0x** [comment]
@ChiTimesChi docs are live [here](https://contracts-interchain.vercel.app)

File: .github/workflows/solidity.yml [hunk 15]
```diff
@@ -26,15 +28,16 @@ jobs:
...
           fetch-depth: 0
-          # if any of these packages use submodules in the future, please uncomment this line
-          # submodules: 'recursive'
+          submodules: 'recursive'
       - uses: dorny/paths-filter@v2
         id: filter_solidity
         with:
           #  make sure to update run-goreleaser when adding a new package here
           # also add to the get-project-id step
+          # TODO: forge doc, get-project-id for contracts-interchain
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply .github/workflows/solidity.yml:37]: sure!

**Pull request closed without merging**